### ISSUE TITLE
Dependency scope

### ISSUE THREAD

**Issue by ZIJ** [diggerhq/digger#791]
Only trigger downstream dependent projects if certain outputs change in the upstream (not any change)

Similar to stack dependencies in Spacelift

From user feedback call today (Alexander F)

**Comment by ZIJ** [comment]
Additional feedback from Francois (NT): "all projects are triggered if modules change, even though changes in certain modules aren't affecting certain projects".

**Comment by ZIJ** [comment]
### Idea for using git as a store for dependencies / TFVars

In digger.yml, each project can define `exports` and `imports` lists, like this:

```
projects:
  - name: projA
    dir: ./proj-a
    exports: ["vpcId", "subnetId"]
  - name: projB
    dir: ./proj-b
    imports: ["projA.vpcId", "projA.subnetId"]
```

On every successful apply of a project, all its exports are pushed to git as `outputs.tfvars` file, into either a separate "infra state" repo or a special folder like `.digger`. Folder structure matches digger.yml, like this:

```
.digger
   ├- ProjA
       ├- outputs.tfvars
   ├- ProjB
       ├- outputs.tfvars
```
This ways multi-stage applies for every change can be split in "layers" (effectively levels of the dependency tree):
- Layer 0: A code change triggers an apply of 1 or more porjects. They run in parallel and each produce outputs stored in git
- Layer 1: All projects that consume outputs of layer 0 as inputs
- ...

User may choose to not proceed with Layer 1 after Layer 0 has been applied, or only apply some of the projects by hand and not all.

**Comment by motatoes** [comment]
Good concept but I don't think outputs or inputs live in git, they are already stored by terraform in state of every project and can be fetched out and passed to state on demand, the state is independent of how many jobs or threads exist and when it comes to reading it it is also safe with concurrent reads. That's how terragrunt does input/output management, and I think we should follow a similar route if we were to build our own dependency management between modules. 

We may wish to cache these dependencies if we are scaling with jobs however and for that we will need either a cloud based cache which can be part of the orchestrator, or some other secret store to rely on caching these values.

**Comment by ZIJ** [comment]
More user feedback (M. U. 13.02.2024):

_is there a way to handle dependant projects where plans for the second project wait until applies are done on the first project? i just (successfully) tried out project dependencies for the first time in that sample PR I provided previously and it seems like plans run right after each other, when I would think (but I might be wrong...) for most cases the second project's plan depends on resources/outputs from the first project's apply..._

_what I want to do (going off your example config):_
_1. have digger run a plan on core, wait for apply request if there are expected changes on core
2. have me, as user, run digger apply to apply changes on core
3. once successful, have digger run a plan on platform, now that core is up to date
4. have me run digger apply to apply changes on platform
merge_

_if two projects are sharing data (e.g. in the above PR via remote_state), the current digger behaviour would result the plan for platform failing when it tries to reference a value that doesn't yet exist (because so far it's only appeared in a plan on core and is not actually in the state for core yet)_

_in a sense it kind of is supported but in a hacky way. you'd just let the second project plan's job fail (or cancel it ahead of time knowing it'll fail) and run digger apply -p core, then a manual digger plan again and then finally digger apply (and probably also limit these to -p platform while we're at it)_

**Comment by vedant-rox** [comment]
Opened a first PR for this here: #2638

This is intentionally a safe first slice of #791:
- opt-in dependency_file_triggers based on real Terraform/OpenTofu import graphs
- manual Terragrunt dependency graph support for trigger discovery
- inferred project dependency edges when imported files map to another Digger project
- default behavior remains unchanged

It improves changed-file targeting and reduces duplicated depends_on config, but it does not yet implement output-aware downstream triggering or apply-before-plan sequencing. Happy to take that on as a follow-up if this scope looks right.

**Comment by vedant-rox** [comment]
The first reviewable split for this work is now in [#2639](https://github.com/diggerhq/digger/pull/2639).

I split the original broader PR [#2638](https://github.com/diggerhq/digger/pull/2638) after review feedback asked for smaller PRs that are easier to validate in GitHub Actions.

Current split:
- [#2639](https://github.com/diggerhq/digger/pull/2639): Terraform/OpenTofu-only `dependency_file_triggers`, recursive local import discovery, inferred project dependency edges, and hard-mode propagation constrained to matching imported-file changes
- [#2638](https://github.com/diggerhq/digger/pull/2638): temporary staging branch for the Terragrunt-specific follow-up work and hardening

The remaining Terragrunt/output-aware pieces are still follow-up work and are not part of the first reviewable slice.

**Comment by vedant-rox** [comment]
The current first reviewable split for this work is now in [#2640](https://github.com/diggerhq/digger/pull/2640).

I split the earlier PR again after review and follow-up investigation showed the inferred dependency-edge / hard-mode propagation work was still touching GitHub/GitLab event semantics more than I wanted for the first slice.

Current split:
- [#2640](https://github.com/diggerhq/digger/pull/2640): Terraform/OpenTofu matching only — opt-in `dependency_file_triggers`, recursive local module import discovery, generated-project support, validation, docs, and `digger_config` tests
- [#2639](https://github.com/diggerhq/digger/pull/2639): follow-up branch for inferred dependency edges and hard dependency propagation behavior
- [#2638](https://github.com/diggerhq/digger/pull/2638): older superseded broad branch

If maintainers prefer, I can keep iterating on the follow-up branch separately after feedback on the smaller matching-focused slice.
